lib/ts/nextjs/middleware.ts

```diff
@@ -0,0 +1,438 @@
+    return redirect(redirectUrl.toString());
+}
+
+function redirect(location: string): Response {
+    logDebugMessage(`Redirecting to: ${location}`);
+    return new Response(null, {
+        status: 302,
+        headers: {
+            Location: location,
+        },
+    });
+}
+
+function next(): Response {
+    const response = new Response(null, {});
+    response.headers.set("x-middleware-next", "1");
+    return response;
+}
+
+async function fetchNewTokens(request: Request): Promise<{
+    headers: [string, string][];
+    cookies: string[];
+}> {
+    const refreshApiURL = new URL(
+        `${AppInfo.apiBasePath.getAsStringDangerous()}/session/refresh`,
+        AppInfo.apiDomain.getAsStringDangerous()
+    );
+    const cookieRefreshToken = getCookie(request, REFRESH_TOKEN_COOKIE_NAME);
+    const headerRefreshToken = getCookie(request, REFRESH_TOKEN_HEADER_NAME);
+    const refreshRequestHeaders = new Headers();
+    refreshRequestHeaders.append("Content-Type", "application/json");
+    if (cookieRefreshToken) {
+        refreshRequestHeaders.append("Cookie", `${REFRESH_TOKEN_COOKIE_NAME}=${cookieRefreshToken}`);
+    } else if (headerRefreshToken) {
+        refreshRequestHeaders.append("Authorization", `Bearer ${headerRefreshToken}`);
+    }
+
+    const refreshResponse = await fetch(refreshApiURL, {
+        method: "POST",
+        headers: refreshRequestHeaders,
+        credentials: "include",
+    });
+
+    if (!refreshResponse.ok) {
+        throw new Error(`Refresh request returned an invalid status code: ${refreshResponse.status}`);
+    }
+
+    logDebugMessage("Session refresh request completed");
+    const cookies: string[] = [];
+    const tokens = {
+        accessToken: refreshResponse.headers.get(ACCESS_TOKEN_HEADER_NAME) || "",
+        refreshToken: refreshResponse.headers.get(REFRESH_TOKEN_HEADER_NAME) || "",
+        frontToken: refreshResponse.headers.get(FRONT_TOKEN_HEADER_NAME) || "",
+        antiCsrfToken: refreshResponse.headers.get(ANTI_CSRF_TOKEN_HEADER_NAME) || "",
+    };
+
+    cookies.push(`${FRONT_TOKEN_COOKIE_NAME}=${tokens.frontToken}; Path=/`);
+    // TOOD: Review the current build target
+    // getSetCookie was added in node 18 and our build target is ES5
+    // This should not a problem here since the function runs in the Vercel edge runtime environment
+    // @ts-expect-error TS(2339): Property 'getSetCookie' does not exist on type 'Headers'.
+    const setCookieHeaders = refreshResponse.headers.getSetCookie();
+    for (const header of setCookieHeaders) {
+        if (header.includes(ACCESS_TOKEN_COOKIE_NAME)) {
+            cookies.push(header);
+        }
+        if (header.includes(REFRESH_TOKEN_COOKIE_NAME)) {
+            cookies.push(header);
+        }
+        if (header.includes(ANTI_CSRF_TOKEN_COOKIE_NAME)) {
+            cookies.push(header);
+        }
+    }
+
+    if (!cookies.some((cookie) => cookie.includes(ACCESS_TOKEN_COOKIE_NAME))) {
```

Comment:
Checking like this feels a bit strange - shouldn't we just check if the original refresh response had the response header for header-based sessions or not?

Reply:
Also, keep in mind that there are configurations (`exposeAccessTokenToFrontendInHeaderBasedAuth`) that'd mean both the cookie and the response header are there.